_posts/2016-04-01-rust-via-its-core-values.markdown

```diff
@@ -0,0 +1,343 @@
+---
+layout: post
+title: "Rust via its Core Values"
+date: 2016-04-01T11:21:05-05:00
+author: "Ian Whitney"
+---
+
+I have a couple of ideas about learning programming languages.
+
+The first is that we do it wrong. I'm sure you've experienced this. You try learning a new language and can't quite see how it all works. Why do I use this syntax here and some other syntax over here? We get frustrated by all the weird bits and go back to a language we know.
+
+I think we do a bad job of learning languages because of how we think about languages. Think back to the last time you discussed a new language. Someone mentioned a new language and then someone else asked about the language's speed, syntax or web framework of choice.
+
+This is similar to how we talk about cars. Hear about that new Ford Bratwurst? How fast is it? Can I drive it through a lake?
+
+When we talk about languages this way, we make them sound interchangeable. Like cars. I know how to drive a Toyota Hamhock, so I should be able to drive a Ford Bratwurst, no problem. Only the speed and the dashboard vary, right?
+
+But imagine what a PHP car would look like, now imagine how different the Lisp car would be. Moving from one to the other involves a lot more than learning which knob controls the heater.
+
+<!--break-->
+
+Cars are interchangeable because they all solve the same problem; and we've settled on what is the best way to solve that problem. But programming languages all solve different problems, and they each express their own philosophy about how to solve that problem.
+
+A language's syntax & speed are expressions of its core values. For example, Ruby famously values Developer Happiness and that value has impacted Ruby's features. Java has placed a huge value in backwards compatibility and that has impacted Java's features.
+
+So, my second idea about learning a programming language is this. We can learn a language better by learning it through its core values. If we can anchor our understanding in *why* a language works the way it does, it will be easier to learn the *how*
+
+With that, let's look at Rust's core values:
+
+Rust defines its three goals as
+
+- Speed
+- Memory Safety
+- Concurrency
+
+Let's set aside concurrency for now and focus on Rust's top two goals. Speed and Memory Safety.
+
+A digression. When Rust says it values "memory safety", it is saying that it will not crash with a segmentation fault. This will be familiar to you if you've worked with C or C++, but if (like me) you've avoided those languages then this may be a new idea. Imagine the following situations
+
+- You try to use the 5th element of a 2-element array
+- You try to call a method on Nil
+- Two functions mutate a variable at the same time, leaving it an uncertain state
+- You forget to free some memory
```

Comment:
Thanks for the clarification. As someone who mostly works with GC'd languages, I found it hard to pick good examples. I'll remove that line.